### PR TITLE
Create Rego rules for OPA

### PR DESCRIPTION
**Pull request by siegfriedweber**
# Description

Create Rego rules for OPA which facilitate the definition of policies

Closes stackabletech/issues#500

The Rego rules are contained in an integration test in the directory `tests/templates/kuttl/opa-authorization/trino_rules/`.

Currently, these rules must be deployed manually:

```shell
kubectl create configmap trino-opa-rules \
    --from-file=trino_rules/actual_permissions.rego \
    --from-file=trino_rules/requested_permissions.rego \
    --from-file=trino_rules/util.rego \
    --from-file=trino_rules/verification.rego

kubectl label configmaps trino-opa-rules \
    opa.stackable.tech/bundle=true
```

The custom policies can be defined in a separate file:

```rego
package trino_policies

import rego.v1

policies := {
    "catalogs": [
        {
            "user": "testuser",
            "catalog": "system",
            "allow": "read-only",
        },
    ],
}
```

This file must also be deployed via a ConfigMap:

```shell
kubectl create configmap trino-opa-policies \
    --from-file=trino_policies.rego

kubectl label configmaps trino-opa-policies \
    opa.stackable.tech/bundle=true
```

There is no documentation for the rules yet. But the rules replicate the File-based access control and its [documentation](https://trino.io/docs/current/security/file-system-access-control.html) can be used instead. All rules are implemented but roles are not matched.

The rules are formatted with `opa fmt`:

```shell
opa fmt --write *.rego
```

The rules can be linted with [Regal](https://github.com/StyraInc/regal):

```shell
regal lint *.rego
```

The unit tests can be run with `opa test` in the directory `tests/templates/kuttl/opa-authorization/trino_rules/`:

```shell
opa test --schema schema/ --bundle .
```

The integration test can be run with:

```shell
scripts/run-tests --operator trino=0.0.0-pr559 --test opa-authorization
```

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [x] (Integration-)Test cases added
- [x] ~~Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).~~
- [ ] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: rust/operator-binary/src/authorization/opa.rs [hunk 1]
```diff
@@ -19,6 +19,8 @@ pub struct TrinoOpaConfig {
     opa_authorizer_name: String,
     non_batched_connection_string: String,
     batched_connection_string: Option<String>,
+    row_filters_connection_string: Option<String>,
+    column_masking_connection_string: Option<String>,
     allow_permission_management_operations: bool,
```
Comment (fhennig) [inline rust/operator-binary/src/authorization/opa.rs:24]: Can you add more documentation to this struct? Are these settings that are set in the trino config (I guess they are) - where are these connection strings pointing to? Specific opa endpoints?

**Comment by siegfriedweber** [comment]
https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/trino-operator-it-custom/118/

File: tests/templates/kuttl/opa-authorization/00-rbac.yaml.j2 [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: use-integration-tests-scc
+rules:
+{% if test_scenario['values']['openshift'] == "true" %}
+  - apiGroups: ["security.openshift.io"]
+    resources: ["securitycontextconstraints"]
+    resourceNames: ["privileged"]
+    verbs: ["use"]
+{% endif %}
```
Comment (fhennig) [inline tests/templates/kuttl/opa-authorization/00-rbac.yaml.j2:12]: Good to have this here, can you also please run the tests on OpenShift to make sure the changes are OpenShift compatible?

File: tests/templates/kuttl/opa-authorization/11-opa-rego.yaml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  - script: |
+      # Trino Rego rules provided by the Stackable Data Platform
+      kubectl create configmap trino-opa-rules \
+        --namespace=$NAMESPACE \
+        --from-file=trino_rules/actual_permissions.rego \
+        --from-file=trino_rules/requested_permissions.rego \
+        --from-file=trino_rules/util.rego \
+        --from-file=trino_rules/verification.rego
+      kubectl label configmaps trino-opa-rules \
+        opa.stackable.tech/bundle=trino \
```
Comment (fhennig) [inline tests/templates/kuttl/opa-authorization/11-opa-rego.yaml:14]:
```suggestion
        opa.stackable.tech/bundle=true \
```

File: tests/templates/kuttl/opa-authorization/11-opa-rego.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        --namespace=$NAMESPACE
+
+      # Trino policies provided by the customer
+      kubectl create configmap trino-opa-policies \
+        --namespace=$NAMESPACE \
+        --from-file=trino_policies.rego
+      kubectl label configmaps trino-opa-policies \
+        opa.stackable.tech/bundle=trino \
```
Comment (fhennig) [inline tests/templates/kuttl/opa-authorization/11-opa-rego.yaml:22]:
```suggestion
        opa.stackable.tech/bundle=true \
```

**Review by fhennig — CHANGES_REQUESTED** [review]
just a couple of things, but looks very nice in general!

File: tests/templates/kuttl/opa-authorization/11-opa-rego.yaml (hunk 3 above)
Comment (siegfriedweber) [reply tests/templates/kuttl/opa-authorization/11-opa-rego.yaml]: changed in ed98c309b897a4ae2f098a20ba9c77304b848e2a

File: tests/templates/kuttl/opa-authorization/00-rbac.yaml.j2 (hunk 2 above)
Comment (siegfriedweber) [reply tests/templates/kuttl/opa-authorization/00-rbac.yaml.j2:12]: fixed in 2f6057c88ed1efcdc080df433169fe7d34d77278 and successfully tested on OpenShift 4.12

**Review by fhennig — CHANGES_REQUESTED** [review]
Can you update the changelog?

Also, the Rego rules are barely commented. I don't know what to request specifically, but Rego is not a very well known language and it would be good to have some commentary on the general code flow or setup of the 3 different files. Maybe just a header comment in each file where the "entrypoints" are or which other files call this one? 

I suspect that 'verification.rego' contains the main 'allow' entrypoint, and actual_permissions calculates the permissions based on the policies, and 'requested_permissions' based on the input? and verification calls both of them?

File: tests/templates/kuttl/opa-authorization/trino_rules/util.rego [hunk 5]
```diff
@@ -1,7 +1,18 @@
+# METADATA
+# description: Utility package which extends the built-in functions
 package util
 
 import rego.v1
 
+# METADATA
+# description: |
+#   Matches the entire string againts a regular expression.
```
Comment (fhennig) [inline tests/templates/kuttl/opa-authorization/trino_rules/util.rego:9]:
```suggestion
#   Matches the entire string against a regular expression.
```

**Review by fhennig — CHANGES_REQUESTED** [review]

File: tests/templates/kuttl/opa-authorization/trino_rules/util.rego [hunk 6]
```diff
@@ -6,7 +6,7 @@ import rego.v1
 
 # METADATA
 # description: |
-#   Matches the entire string againts a regular expression.
+#   Matches the entire string against a regular expression.
 #
 #   pattern (string)  regular expression
 #   value (string)    value to match againts pattern
```
Comment (fhennig) [inline tests/templates/kuttl/opa-authorization/trino_rules/util.rego]:
```suggestion
#   value (string)    value to match against pattern
```

**Review by fhennig — CHANGES_REQUESTED** [review]

**Review by fhennig — APPROVED** [review]
Lovely!

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Added
+
+- Support row filters and column masks in Rego rules ([#559]).
```
Comment (sbernauer) [inline CHANGELOG.md:9]: I think we should mark this as breaking, as users of 442 need to adopt their rego rules when updating, WDYT?

Comment (fhennig) [reply CHANGELOG.md:9]: @siegfriedweber

Comment (siegfriedweber) [reply CHANGELOG.md:9]: This is actually not breaking. If an OPA rule is queried which is not defined then it is "undefined". This is the same behavior if the rule is defined but the condition is not met.

The response from OPA is in both cases:

```HTTP
HTTP/1.1 200 OK
Content-Type: application/json
Vary: Accept-Encoding
Date: Wed, 27 Mar 2024 10:13:44 GMT
Content-Length: 55

{"decision_id":"facecbef-f851-40e7-afce-c46cf45b9f2c"}
```

That is, if the entry points for row filters and column mask are not defined then the behavior is the same if these entry points are defined but no row filters and column masks are defined in the policies.

Comment (sbernauer) [reply CHANGELOG.md:9]: Even better, awesome! Thanks for looking it up

**Pull request merged**